### PR TITLE
Add humidity support to climate

### LINKED ISSUE #2306: Add Humidity to Climate Component
**Describe the problem you have/What new integration you would like**

The climate component currently lacks the ability to set current humidity % or set target humidity %.

The climate component in home assistant already supports both. All esphome would have to do is provide these settings optionally for those with humidifiers or dehumidifiers.

**Please describe your use case for this integration and alternatives you've tried:**

The alternative is to use a separate sensor for current humidity and a number component for target humidity which is not ideal for UX.

**Additional context**

### PR DESCRIPTION
**Pull request by kroimon**
# What does this implement/fix?

Adds humidity ~~and auxiliary heater~~ support to the climate component.

Related PRs:
- esphome/aioesphomeapi#631
- home-assistant/core#103807

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):**
Closes esphome/feature-requests#2306

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3359

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
How exactly would auxiliary heat be used? (a doc example could help)

I guess Thermostat, Bang-Bang, PID Climate should be extended with aux heat and humidity capability...?

Typically a templatable switch/binary output could control the aux heater.

(reposted comment from the wrong repo)

**Comment by kroimon** [comment]
> How exactly would auxiliary heat be used? (a doc example could help)
> Typically a templatable switch/binary output could control the aux heater.

This PR just adds the feature to the base climate platform component to align it with Home Assistant. Specific climate components would later need to implement it. It is then up to them how to react to a request by the user to turn on/off aux_heat.
For example, a heat pump could turn on an additional electric heater if outside temperatures are too low for it to properly operate (sometimes also called emergency heat).
Basically, aux_heat is just an additional command users can pass directly from the HA climate GUI to a component implementing a climate device.

> I guess Thermostat, Bang-Bang, PID Climate should be extended with aux heat and humidity capability...?

I agree, but not for this PR. This focuses on the base implementation to keep it reviewable. Specific climate components can then implement the new features at a later stage.

**Comment by nagyrobi** [comment]
Having these mentioned in the docs, but not having at least one single real component which can make use of them will generate confusion and support demand, we will have to explain it one by one each time somebody asks why is there but not usable.

I suggest to either:
- add them to at least one climate which can really use them (as above mentioned)
- or make a temporary note in te doc that it's work in progress
- or stay publically undocumented (added to doc, but commented out)

**Comment by kroimon** [comment]
Just to leave a quick update: I am going to add support for at least current humidity and auxiliary heaters to at least one of the above mentioned components as soon as I find time for it.

If possible, I would love to get some additional feedback on this PR to make sure I don't waste time implementing it for these components if this main PR already needs major changes.

**Comment by kbx81** [comment]
> > I guess Thermostat, Bang-Bang, PID Climate should be extended with aux heat and humidity capability...?
> 
> I agree, but not for this PR. This focuses on the base implementation to keep it reviewable. Specific climate components can then implement the new features at a later stage.

I also agree -- let's not make this PR any bigger and keep it as just a base implementation to support the additional functionality. Bigger PRs are more difficult to review.

**Comment by kroimon** [comment]
I implemented current humidity support to the `bang_bang`, `pid` and `thermostat` climate controllers.
Adding target humidity or auxiliary heater support to those is a bigger task that feels out of scope for now and requires more thought, design work and probably some discussion on how to do it best, if at all.

I plan on also adding current humidity support to the `uponor_smatrix` component from #5769 as soon as this here has been merged.

I declare this **Ready for Review™**.

**Review by kbx81 — DISMISSED** [review]
LGTM and thanks for working on this. 😄

**Comment by Daniel-dev22** [comment]
Is there an idea of when we might see this feature in a release?

**Comment by kroimon** [comment]
I have removed deprecated auxiliary heater support from this PR and created reverting PRs for API and HA:
- https://github.com/home-assistant/core/pull/104694
- https://github.com/esphome/aioesphomeapi/pull/783

Architecture decision deprecating aux_heat: https://github.com/home-assistant/architecture/discussions/932

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — COMMENTED** [review]
An afterthought while going through the docs, I understand reporting the humidity value, but why would the climate platforms say they support setting target humidity when its never even used?
Until such time as the code is written on ESPHome side to actually use the target humidity, shouldnt that flag stay as false and not be based on whether or not you have a humidity sensor attached.

**Comment by kroimon** [comment]
> An afterthought while going through the docs, I understand reporting the humidity value, but why would the climate platforms say they support setting target humidity when its never even used? Until such time as the code is written on ESPHome side to actually use the target humidity, shouldnt that flag stay as false and not be based on whether or not you have a humidity sensor attached.

I'm not sure I understand. The `supports_target_humidity` flag is never currently set, only the `supports_current_humidity` flag is used where appropriate. Or do I have a typo somewhere that I missed?

**Comment by bdraco** [comment]
AFAICT `set_supports_target_humidity` is never called?

**Comment by jesserockz** [comment]
lol, I am an idiot and misread it

File: esphome/components/mqtt/mqtt_climate.cpp [hunk 1]
```diff
@@ -57,6 +60,11 @@ void MQTTClimateComponent::send_discovery(JsonObject root, mqtt::SendDiscoveryCo
     root[MQTT_TEMPERATURE_STATE_TOPIC] = this->get_target_temperature_state_topic();
   }
 
+  // target_humidity_command_topic
+  root[MQTT_TARGET_HUMIDITY_COMMAND_TOPIC] = this->get_target_humidity_command_topic();
+  // target_humidity_state_topic
+  root[MQTT_TARGET_HUMIDITY_STATE_TOPIC] = this->get_target_humidity_state_topic();
+
```
Comment (jesserockz) [inline esphome/components/mqtt/mqtt_climate.cpp]:
```suggestion
  if (traits.get_supports_target_humidity()) {
    // target_humidity_command_topic
    root[MQTT_TARGET_HUMIDITY_COMMAND_TOPIC] = this->get_target_humidity_command_topic();
    // target_humidity_state_topic
    root[MQTT_TARGET_HUMIDITY_STATE_TOPIC] = this->get_target_humidity_state_topic();
  }

```

Comment (kroimon) [reply esphome/components/mqtt/mqtt_climate.cpp]: Good catch!

**Review by jesserockz — APPROVED** [review]
